src/raftstore/store/read_queue.rs

```diff
@@ -62,17 +62,48 @@ impl Drop for ReadIndexRequest {
     }
 }
 
-#[derive(Default)]
 pub struct ReadIndexQueue {
     reads: VecDeque<ReadIndexRequest>,
     ready_cnt: usize,
     // How many requests are handled.
     handled_cnt: usize,
     // map[uuid] -> offset in `reads`.
     contexts: HashMap<Uuid, usize>,
+
+    retry_timeout: usize,
+    retry_elapsed: usize,
+    last_retried: usize,
 }
 
 impl ReadIndexQueue {
+    pub fn new(cfg: &Config) -> Self {
+        ReadIndexQueue {
+            reads: VecDeque::default(),
+            ready_cnt: 0,
+            handled_cnt: 0,
+            contexts: HashMap::default(),
+            retry_timeout: cfg.raft_election_timeout_ticks,
+            retry_elapsed: 0,
+            last_retried: 0,
+        }
+    }
+
+    pub fn check_needs_retry(&mut self) -> bool {
+        self.retry_elapsed += 1;
+        if self.retry_elapsed < self.retry_timeout {
+            return false;
+        }
+
+        self.retry_elapsed = 0;
+        let total = self.reads.len();
+        if total == self.ready_cnt || total + self.handled_cnt <= self.last_retried {
+            return false;
+        }
+
+        self.last_retried = total + self.handled_cnt - 1;
```

Comment:
Why minus one?

Reply:
So `last_retried` is an index, not a length.